### ISSUE TITLE
Server should refuse to start if configured version is not semver.org compatible

### ISSUE THREAD

**Issue by nikclayton** [mastodon/mastodon#28843]
### Pitch

It seems as though the server administrator can configure the server's reported version number.

For example, queer.party currently reports the version as "4.1.8+trans_rights-0aef33b-23262162413".

This is not a https://semver.org compatible version number (the underscore is an invalid character), which can cause problems for Mastodon clients. To prevent misconfigurations like that the server should refuse to start, and report an error that allows the server administrator to easily correct the problem.

This check **should not** be ignorable via a configuration setting, command line flag, or other mechanism. If the server administrator wants to break this compatibility they should be required to edit the code to do so, with clear warnings as to the impact that will have on clients that connect to their server.

### Motivation

Clients need to reliably distinguish between different versions of the Mastodon server, because different versions introduce new functionality or change how existing functionality works, and clients need to be able to adapt to that.

Being able to reliably parse the server's version number is an important part of that.

> **Note**: This has *nothing* to do with how you decide to change the version number, or whether you follow https://semver.org's recommendations for when to bump the major/minor/patch versions (which is what derailed https://github.com/mastodon/mastodon/issues/8297). This is *only* about ensuring that the version number presented by the server can be parsed reliably.

**Comment by deviant** [comment]
Crosslinking [my response](https://iceshrimp.dev/iceshrimp/iceshrimp-rewrite/issues/1#issuecomment-3518) to the same issue as submitted to another software project. I think that using a generic version parsing library [rather than](https://en.wikipedia.org/wiki/Robustness_principle) using a semver library to parse the version (as the author is in their project, not mentioned above) is more sensible, since it cannot be assumed that all Mastodon instances will want to use semver, nor that all forks will comply with this and retain such an assertion. Breaking any instance with an unusual version number seems like especially bad UX.

**Comment by Joshix-1** [comment]
I would just patch that check out. So please don't

**Comment by nikclayton** [comment]
@Joshix-1 Why would you deliberately patch out functionality that would improve the interoperability between your server and different clients?

That's a not a rhetorical question -- I am genuinely interested in understanding why you would choose to make it harder for clients to provide functionality for your users.

@deviant There is no generic version parsing library. The current set of version formats used by Mastodon-like servers is a hot mess; https://github.com/pachli/pachli-android/blob/main/core/network/src/test/resources/server-versions.json5 is a snapshot from the fediverse.observer data (includes servers that do not implement the Mastodon API). Fixing that won't happen overnight, but the work has to start somewhere.

**Comment by Joshix-1** [comment]
Can't you just compare the version strings numerically ("9" less than "17")? Why do you need to parse it? I don't see why you would need to rely on the version being full semver compatible (servers could always just not honor that)?

A "_" in a version number should never break a client.

**Comment by nikclayton** [comment]
@Joshix-1 

> Can't you just compare the version strings numerically ("9" less than "17")?

No.

For example, "4.1.0" is a higher version than "4.1.0-beta", which is a higher version than "4.1.0-alpha". But "4.1.0-nightly" is a higher version than "4.1.0" (maybe -- perhaps this is the nightly release candidate for what will become 4.1.0?).

Or consider these, from Akkoma:

- 3.10.3-0-g368b22f
- 3.10.3-0-g43c5fd5
- 3.10.3-0-g6fb91d7
- 3.10.3-0-g98f0820

Same base version, those are probably git commit IDs with a "g" prefix, but those IDs don't encode anything about which one represents a newer version.

But that doesn't answer my question -- why would you choose to make it harder for clients to provide functionality for your users?

**Comment by ThisIsMissEm** [comment]
This seems reasonable, I think?

**Comment by nikclayton** [comment]
@kescherCode 

> I think clients could simply parse these numbers and that's about it, ignore the rest

You're arguing that clients should consider "4.2.5" and "4.2.5-alpha.1" to be the same version, correct?

Can you envision circumstances where that would not be the right thing to do?

**Comment by kescherCode** [comment]
@nikclayton I don't see any issue with only parsing the start of a version string. The only reason the suffix exists now is due to the update check, because the update-check API (which is not open-source currently, so I have to take a wild guess) seems to take this suffix into account due to nightly version deployments. Before that, even nightly versions of Mastodon only ever adhered to semver.

As such, I propose: If you want Mastodon to terminate on invalid version, only do so if the start isn't semver. Anything after a valid semver shall be ignored by clients as it currently already is the case in most.

**Comment by nikclayton** [comment]
@kescherCode Another use case that's just come up given the current ongoing spam wave -- admins want to suspend remote instances that are running versions of the software that are known to have security issues.

As one of them has just commented, "the challenge is that alpha, alpha.1, and alpha.2 all have critical vulnerabilities".

Your "only parsing the start of the version string" suggestion is actively harmful in this case.

**Comment by ThisIsMissEm** [comment]
@nikclayton I'm sure this was mentioned before, but the example version string is technically valid:

> Build metadata MAY be denoted by appending a plus sign and a series of dot separated identifiers immediately following the patch or pre-release version. Identifiers MUST comprise only ASCII alphanumerics and hyphens [0-9A-Za-z-]. Identifiers MUST NOT be empty. Build metadata MUST be ignored when determining version precedence. Thus two versions that differ only in the build metadata, have the same precedence. Examples: 1.0.0-alpha+001, 1.0.0+20130313144700, 1.0.0-beta+exp.sha.5114f85, 1.0.0+21AF26D3----117B344092BD.

https://semver.org/spec/v2.0.0.html#spec-item-10

However, yes, we should encourage forks to change their software name, i.e., `mastodon-glitch` or `mastodon-trans_rights` or whatever. But then you need rules for parsing the name to understand "are these two things the same?"

**Comment by nikclayton** [comment]
@ThisIsMissEm 

> But then you need rules for parsing the name to understand "are these two things the same?"

Yes. That's obviously a non-starter, so shouldn't be encouraged at all. Software name and version fields are technical data with a defined meaning. Using those fields for other purposes, however well intentioned, is not helpful.

The queer.party example in the first post is no longer relevant because I contacted maffsie to let them know about the problem and they agreed to change it.

But there's plenty of other messed up versions out there, c.f., https://github.com/pachli/pachli-android/blob/main/core/network/src/test/resources/server-versions.json5

**Comment by ThisIsMissEm** [comment]
@nikclayton I think relying on the software name / version is honestly an anti-pattern, hence introducing: #29191 but we could extend the instance info endpoint with more data about supported features

**Comment by nikclayton** [comment]
@ThisIsMissEm check your Mastodon DMs...
